drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,455 @@
+	K_POLL_SIGNAL_INITIALIZER(cmd_done_signal);
+static atomic_t cesq_subscribers;
+
+struct at_cmd_fifo_item {
+	void *fifo_reserved;
+	struct at_cmd_msg cmd;
+};
+
+static void dispatch_handler(struct k_work *work)
+{
+	struct at_cmd_client *client =
+		CONTAINER_OF(work, struct at_cmd_client, work);
+
+	client->handler(client->cmd, client->cmd_len,
+			client->response, client->response_len);
+}
+
+static u32_t client_id_create(void)
+{
+	static atomic_t client_id_counter;
+
+	atomic_inc(&client_id_counter);
+
+	return (u32_t)client_id_counter;
+}
+
+static size_t client_idx_from_id(u32_t id)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(clients); i++) {
+		if (id == clients[i]->id) {
+			return i;
+		}
+	}
+
+	return INVALID_CLIENT_IDX;
+}
+
+static struct at_cmd_client *client_add(size_t client_idx,
+					at_cmd_handler_t handler)
+{
+	struct at_cmd_client *client;
+
+	client = k_malloc(sizeof(struct at_cmd_client));
+	if (client == NULL) {
+		LOG_ERR("Could not allocate memory for client.");
+		return NULL;
+	}
+
+	clients[client_idx] = client;
+	client->id = client_id_create();
+	client->handler = handler;
+	client->cmd_pending = false;
+
+	k_work_init(&client->work, dispatch_handler);
+
+	return client;
+}
+
+struct at_cmd_client *at_cmd_client_init(at_cmd_handler_t handler)
+{
+	struct at_cmd_client *client = NULL;
+
+	__ASSERT_NO_MSG(handler);
+
+	k_mutex_lock(&clients_mutex, K_FOREVER);
+
+	for (size_t i = 0; i < ARRAY_SIZE(clients); i++) {
+		if (clients[i]) {
+			continue;
+		}
+
+		client = client_add(i, handler);
+		break;
+	}
+
+	k_mutex_unlock(&clients_mutex);
+
+	if (!client) {
+		LOG_ERR("Could not register AT command client.");
+	}
+
+	return client;
+}
+
+static void client_free(size_t client_idx)
+{
+	k_free(&clients[client_idx]);
+	clients[client_idx] = NULL;
+}
+
+void at_cmd_client_uninit(struct at_cmd_client *client)
+{
+	size_t client_idx;
+
+	__ASSERT_NO_MSG(client);
+
+	k_mutex_lock(&clients_mutex, K_FOREVER);
+
+	client_idx = client_idx_from_id(client->id);
+	if (client_idx == INVALID_CLIENT_IDX) {
+		LOG_ERR("Client could not be removed, client ID not found.");
+	} else {
+		client_free(client_idx);
+	}
+
+	k_mutex_unlock(&clients_mutex);
+}
+
+int at_cmd_send(struct at_cmd_client *client, struct at_cmd_msg *cmd)
+{
+	struct at_cmd_fifo_item *fifo_cmd_data;
+	char *buf;
+
+	if (cmd->len > CMD_MAX_LEN) {
+		LOG_ERR("AT command length exceeds maximum allowed length.");
+		return -EMSGSIZE;
+	}
+
+	/* Allocate memory for the command struct. */
+	fifo_cmd_data = k_malloc(sizeof(struct at_cmd_fifo_item));
+	if (fifo_cmd_data == NULL) {
+		LOG_ERR("Memory could not be allocated for the command.");
+		return -ENOMEM;
+	}
+
+	/* Allocate memory for the comamnd buffer. */
+	buf = k_malloc(cmd->len);
+	if (buf == NULL) {
+		k_free(fifo_cmd_data);
+		LOG_ERR("Memory could not be allocated for command buffer.");
+		return -ENOMEM;
+	}
+
+	fifo_cmd_data->cmd.client = client;
+	fifo_cmd_data->cmd.timestamp = k_uptime_get_32();
```

Comment:
Not used for anything now, but there was initially a wish for it to be used in the future for caching of AT command responses. The thought is that clients can set a maximum acceptable "age" of an AT command reponse. If the cached response is within the acceptable range, the command response can be the cached one instead of issuing a new command to the modem. Can be used for IP address, operator or something that usually doesn't change often.

Reply:
In this case caching will only increase the complexity without adding much value. I don't think the data traffic going over this socket is high enough.

Reply:
Yeah, you're probably right. Can be added in the future if needed.